clients/rospy/src/rospy/impl/tcpros_pubsub.py

```diff
@@ -386,15 +386,17 @@ def __init__(self, connection, queue_size):
 
         self._lock = threading.Lock()
         self._cond_data_available = threading.Condition(self._lock)
-        self._connection.set_cleanup_callback(self._closed_connection_callback)
         self._queue = []
         self._error = None
 
         self._thread = threading.Thread(target=self._run)
         self._thread.start()
 
+        self._connection.add_cleanup_callback(self._closed_connection_callback)
+
     def _closed_connection_callback(self, connection):
-        self._cond_data_available.notify()
+        with self._lock:
```

Comment:
This would imply that this code path was either never triggered or didn't work before?